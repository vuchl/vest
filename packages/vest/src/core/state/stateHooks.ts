import defaultTo from 'defaultTo';
import type { TStateHandlerReturn } from 'vast';

import VestTest from 'VestTest';
import type { TStateRef } from 'createStateRef';
import ctx from 'ctx';
import type { TDraftResult } from 'produceDraft';

export function useSuiteId(): TStateHandlerReturn<string> {
  return useStateRef().suiteId();
}
export function useTestCallbacks(): TStateHandlerReturn<{
  fieldCallbacks: Record<string, ((res: TDraftResult) => void)[]>;
  doneCallbacks: ((res: TDraftResult) => void)[];
}> {
  return useStateRef().testCallbacks();
}
export function useOptionalFields(): TStateHandlerReturn<
  Record<string, boolean>
> {
  return useStateRef().optionalFields();
}
export function useStateRef(): Exclude<TStateRef, void> {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return ctx.useX().stateRef!; // I should revisit this
}

export function useTestObjects(): TStateHandlerReturn<VestTest[]> {
  return useStateRef().testObjects();
}

export function usePrevTestObjects(): TStateHandlerReturn<VestTest[]> {
  return useStateRef().prevTestObjects();
}

export function useTestAtCursor(initialValue: VestTest): VestTest {
  const [cursorAt] = useCursorAt();
  const [prevTestObjects] = usePrevTestObjects();

  const nextTest = defaultTo(prevTestObjects[cursorAt], initialValue);
  useSetTestAtCursor(nextTest);

  return nextTest;
}

export function useSetTestAtCursor(testObject: VestTest): void {
  const [cursorAt] = useCursorAt();
  const [testObjects, setTestObjects] = useTestObjects();

  if (testObject === testObjects[cursorAt]) {
    return;
  }

  setTestObjects((testObjects: VestTest[]) => {
    const newTestsOrder = testObjects.slice(0);
    newTestsOrder[cursorAt] = testObject;
    return newTestsOrder;
  });
}

export function useSetNextCursorAt(): void {
  const [, setCursorAt] = useCursorAt();

  setCursorAt((cursorAt: number) => cursorAt + 1);
}

export function useRefreshTestObjects(): void {
  const [, setTestObjects] = useTestObjects();

  setTestObjects(testObjects => testObjects.slice(0));
}

export function useCursorAt(): TStateHandlerReturn<number> {
  return useStateRef().testObjectsCursor();
}

export function isOptionalField(fieldName: string): boolean {
  const [optionalFields] = useOptionalFields();

  return !!optionalFields[fieldName];
}

export function useAllIncomplete(): VestTest[] {
  const [testObjects] = useTestObjects();

  // TODO: CACHE?
  return testObjects.filter(testObject => testObject.isPending());
}
