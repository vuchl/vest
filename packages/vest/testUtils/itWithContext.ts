import optionalFunctionValue from 'optionalFunctionValue';

import runCreateRef from './runCreateRef';

import context from 'ctx';

export default function itWithContext(
  str: string,
  cb: () => void,
  getCTX?: () => Record<string, any>
): void {
  return it(str, () =>
    context.run(
      optionalFunctionValue(getCTX) ?? { stateRef: runCreateRef() },
      cb
    )
  );
}
