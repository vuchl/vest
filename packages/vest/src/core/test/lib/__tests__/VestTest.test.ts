import wait from 'wait';

import itWithContext from '../../../../../testUtils/itWithContext';

import VestTest from 'VestTest';
import { useAllIncomplete, useTestObjects } from 'stateHooks';
import * as vest from 'vest';

const fieldName = 'unicycle';
const message = 'I am Root.';

describe('VestTest', () => {
  let testObject;

  beforeEach(() => {
    testObject = new VestTest(fieldName, jest.fn(), {
      message,
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  it('Should have a unique id', () => {
    Array.from(
      { length: 100 },
      () => new VestTest(fieldName, jest.fn(), { message })
    ).reduce((existing, { id }) => {
      expect(existing[id]).toBeUndefined();
      existing[id] = true;
      return existing;
    }, {});
  });

  describe('testObject.warn', () => {
    it('Should set `.warns` to true', () => {
      expect(testObject.warns).toBe(false);
      testObject.warn();
      expect(testObject.warns).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      jest.resetModules();

      const VestTest = require('VestTest').default; // eslint-disable-line @typescript-eslint/no-var-requires
      testObject = new VestTest(fieldName, jest.fn(), { message });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should set status to failed', () => {
      expect(testObject.status).not.toBe('FAILED');
      testObject.fail();
      expect(testObject.status).toBe('FAILED');
    });
  });

  describe('testObject.valueOf', () => {
    test('When test did not fail', () => {
      expect(testObject.valueOf()).toBe(true);
    });

    test('When test failed', () => {
      testObject.fail();
      expect(testObject.valueOf()).toBe(false);
    });
  });

  describe('testObject.cancel', () => {
    it('Should remove a testObject from the state', () => {
      return new Promise<void>(done => {
        let testObject: VestTest;
        const suite = vest.create(() => {
          testObject = vest.test('f1', async () => {
            await wait(100);
          });
          vest.test('f2', async () => {
            await wait(100);
          });
          testObject.cancel();
        });
        suite();
        expect(suite.get().tests.f1).toBeUndefined();
        expect(suite.get().tests.f2).toBeDefined();
        done();
      });
    });

    itWithContext('Should be removed from the list of incomplete tests', () => {
      const [, setTestObjects] = useTestObjects();
      setTestObjects(testObjects => testObjects.concat(testObject));
      testObject.setPending();
      {
        const allIncomplete = useAllIncomplete();

        expect(allIncomplete).toEqual(expect.arrayContaining([testObject]));
      }
      testObject.cancel();
      {
        const allIncomplete = useAllIncomplete();
        expect(allIncomplete).toEqual(expect.not.arrayContaining([testObject]));
      }
    });
  });
});
