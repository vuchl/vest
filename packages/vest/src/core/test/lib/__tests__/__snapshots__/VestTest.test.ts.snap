// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VestTest TestObject constructor 1`] = `
VestTest {
  "fieldName": "unicycle",
  "id": "0",
  "message": "I am Root.",
  "status": "UNTESTED",
  "testFn": [MockFunction],
  "warns": false,
}
`;

exports[`VestTest testObject.warn Should set \`.warns\` to true 1`] = `
VestTest {
  "fieldName": "unicycle",
  "id": "102",
  "message": "I am Root.",
  "status": "UNTESTED",
  "testFn": [MockFunction],
  "warns": true,
}
`;
